### ISSUE TITLE
P(erfect) curvess are used like bezier in older maps

### ISSUE THREAD

**Issue by llllllllll** [llllllllll/slider#13]
while parsing: https://osu.ppy.sh/b/1018661
```
Traceback (most recent call last):
  File "/home/combine/combine/submodules/slider/slider/library.py", line 218, in lookup_by_id
    return self._read_beatmap(self, beatmap_id=beatmap_id)
  File "/home/combine/combine/submodules/slider/slider/library.py", line 191, in _raw_read_beatmap
    return Beatmap.from_path(self._cache[f'id:{beatmap_id}'])
  File "/home/combine/combine/submodules/slider/slider/library.py", line 31, in __getitem__
    return d[key]
KeyError: 'id:1018661'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/combine/combine/combine/irc.py", line 91, in _handle_target
    return self.message_handler(self, user, channel, msg.strip())
  File "/home/combine/combine/combine/handler.py", line 107, in __call__
    f(self, client, user, data)
  File "/home/combine/combine/combine/handler.py", line 341, in wrapper
    f(*args, **kwargs)
  File "/home/combine/combine/combine/handler.py", line 380, in recommend
    for n, candidate in enumerate(self._candidates):
  File "/home/combine/combine/combine/utils.py", line 38, in __next__
    return next(self._iterator)
  File "/home/combine/combine/combine/handler.py", line 203, in _gen_candidates
    len(candidate.beatmap(save=True).hit_objects) >= 2):
  File "/home/combine/combine/submodules/slider/slider/client.py", line 80, in _beatmap
    save=save,
  File "/home/combine/combine/submodules/slider/slider/library.py", line 222, in lookup_by_id
    return self.download(beatmap_id, save=save)
  File "/home/combine/combine/submodules/slider/slider/library.py", line 335, in download
    beatmap = Beatmap.parse(data.decode('utf-8-sig'))
  File "/home/combine/combine/submodules/slider/slider/beatmap.py", line 1701, in parse
    groups['HitObjects'],
  File "/home/combine/combine/submodules/slider/slider/beatmap.py", line 375, in parse
    return parse(Position(x, y), time, hitsound, rest)
  File "/home/combine/combine/submodules/slider/slider/beatmap.py", line 702, in _parse
    Curve.from_kind_and_points(slider_type, points, pixel_length),
  File "/home/combine/combine/submodules/slider/slider/curve.py", line 23, in from_kind_and_points
    return subcls(points, req_length)
  File "/home/combine/combine/submodules/slider/slider/curve.py", line 134, in __init__
    self._center = center = get_center(*self.points)
  File "/home/combine/combine/submodules/slider/slider/curve.py", line 212, in get_center
    ((p2.x - p1.x) * (p3.y - p2.y))
ZeroDivisionError: division by zero
```
cc @blimmo

**Comment by llllllllll** [comment]
I think I figured this out. There are two problems here:

Old maps use P with more than 3 points or sometimes 2 points. I'm this case, osu treats it as a Bezier curve.

Even after that change I still got issues. This is because there is no circle that passes through three collinear points. If this occurs, osu handles it as a Bezier curve which is just a line. I have a linear algebra solution for get center which removes the branches and should be less code.

**Comment by blimmo** [comment]
oopsies

**Comment by llllllllll** [comment]
as another note: the osu! source calls the `P` curves `PerfectCurve`: https://github.com/ppy/osu/blob/master/osu.Game/Rulesets/Objects/Types/CurveType.cs, we should update the name of our class.

**Comment by blimmo** [comment]
I think this was closed by #21
